clang/test/SemaSYCL/check-direct-attribute-propagation.cpp

```diff
@@ -0,0 +1,289 @@
+// RUN: %clang_cc1 -fsycl-is-device -internal-isystem %S/Inputs -Wno-sycl-2017-compat -sycl-std=2020 -ast-dump %s | FileCheck %s
+
+// Tests for AST of [[intel::scheduler_target_fmax_mhz()]], [[intel::num_simd_work_items()]],
+// [[intel::no_global_work_offset()]], [[intel::max_global_work_dim()]], [[intel::sycl_explicit_simd]],
+// [[sycl::reqd_sub_group_size()]], [[sycl::reqd_work_group_size()]], [[intel::kernel_args_restrict]], and
+// [[intel::max_work_group_size()]] function attributes in SYCL 2020.
+
+#include "sycl.hpp"
+
+sycl::queue deviceQueue;
+
+struct FuncObj {
+  [[intel::sycl_explicit_simd]] void operator()() const {}
+};
+
+struct FuncObj1 {
+  [[intel::no_global_work_offset(1)]] void operator()() const {}
+};
+
+struct FuncObj2 {
+  [[intel::scheduler_target_fmax_mhz(10)]] void operator()() const {}
+};
+
+struct FuncObj3 {
+  [[intel::max_work_group_size(2, 2, 2)]] void operator()() const {}
+};
+
+struct FuncObj4 {
+  [[sycl::reqd_work_group_size(2, 2, 2)]] void operator()() const {}
+};
+
+struct FuncObj5 {
+  [[intel::num_simd_work_items(8)]] void operator()() const {}
+};
+
+struct FuncObj6 {
+  [[intel::kernel_args_restrict]] void operator()() const {}
+};
+
+struct FuncObj7 {
+  [[intel::max_global_work_dim(1)]] void operator()() const {}
+};
+
+[[intel::sycl_explicit_simd]] void func() {}
+
+[[intel::no_global_work_offset(1)]] void func1() {}
+
+[[intel::scheduler_target_fmax_mhz(2)]] void func2() {}
+
+[[intel::max_work_group_size(1, 1, 1)]] void func3() {}
+
+[[sycl::reqd_work_group_size(1, 1, 1)]] void func4() {}
+
+[[intel::num_simd_work_items(5)]] void func5() {}
+
+[[intel::kernel_args_restrict]] void func6() {}
+
+[[intel::max_global_work_dim(0)]] void func7() {}
+
+[[sycl::reqd_sub_group_size(4)]] void func8() {}
+
+class Functor {
+public:
+  void operator()() const {
+    func8();
+  }
+};
+
+class Functor1 {
+public:
+  [[sycl::reqd_sub_group_size(12)]] void operator()() const {}
+};
+
+int main() {
+  deviceQueue.submit([&](sycl::handler &h) {
+    // CHECK:       FunctionDecl {{.*}}test_kernel1
+    // CHECK:       SYCLSimdAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLKernelAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLSimdAttr
+    h.single_task<class test_kernel1>(
+        FuncObj());
+    // CHECK:       FunctionDecl {{.*}}test_kernel2
+    // CHECK:       SYCLSimdAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLKernelAttr {{.*}} Implicit
+    // CHECK-NEXT:  SYCLSimdAttr
+    h.single_task<class test_kernel2>(
+        []() [[intel::sycl_explicit_simd]]{});
+
+    // Test attribute is not propagated.
+    // CHECK:      FunctionDecl {{.*}}test_kernel3
+    // CHECK:      SYCLSimdAttr {{.*}} Implicit
+    // CHECK-NEXT: SYCLKernelAttr {{.*}} Implicit
+    // CHECK-NEXT: SYCLSimdAttr {{.*}}
+    // CHECK-NOT:  SYCLSimdAttr
+    h.single_task<class test_kernel3>(
+        []() [[intel::sycl_explicit_simd]] { func(); });
```

Comment:
FYI, The error will be generated if we do not pass [[intel::sycl_explicit_simd]]here:

error: SYCL kernel without 'sycl_explicit_simd' attribute can't call a function with this attribute